File: go/internal/realtime/sdp_filter.go
package realtime

import "strings"

func normalizeOffer(sdp string) string {
	var canonicalUfrag, canonicalPwd string
	haveUfrag, havePwd := false, false

	var b strings.Builder
	b.Grow(len(sdp))

	for _, line := range splitInclusive(sdp, '\n') {
		trimmed := strings.TrimRight(line, "\r\n")
		switch {
		case strings.HasPrefix(trimmed, "a=ice-ufrag:"):
			rest := trimmed[len("a=ice-ufrag:"):]
			if !haveUfrag {
				canonicalUfrag = rest
				haveUfrag = true
				b.WriteString(line)
			} else {
				b.WriteString("a=ice-ufrag:")
				b.WriteString(canonicalUfrag)
				b.WriteString(lineTerminator(line))
			}
		case strings.HasPrefix(trimmed, "a=ice-pwd:"):
			rest := trimmed[len("a=ice-pwd:"):]
			if !havePwd {
				canonicalPwd = rest
				havePwd = true
				b.WriteString(line)
			} else {
				b.WriteString("a=ice-pwd:")
				b.WriteString(canonicalPwd)
				b.WriteString(lineTerminator(line))
			}
		default:
			b.WriteString(line)
		}
	}
	return b.String()
}

func splitInclusive(s string, sep byte) []string {
	if s == "" {
		return nil
	}
	var out []string
	start := 0
	for i := 0; i < len(s); i++ {
		if s[i] == sep {
			out = append(out, s[start:i+1])
			start = i + 1
		}
	}
	if start < len(s) {
		out = append(out, s[start:])
	}
	return out
}

func lineTerminator(line string) string {
	if strings.HasSuffix(line, "\r\n") {
		return "\r\n"
	}
	if strings.HasSuffix(line, "\n") {
		return "\n"
	}
	return ""
}

func filterOpusOnly(sdp string) string {
	lines := strings.Split(sdp, "\r\n")

	type mSection struct {
		start, end int
		opusPT     string
	}
	var sections []mSection
	cur := -1
	for i, l := range lines {
		if strings.HasPrefix(l, "m=audio") {
			if cur >= 0 {
				sections[cur].end = i
			}
			sections = append(sections, mSection{start: i, end: len(lines)})
			cur = len(sections) - 1
			continue
		}
		if strings.HasPrefix(l, "m=") && cur >= 0 {
			sections[cur].end = i
			cur = -1
		}
		if cur >= 0 && strings.HasPrefix(l, "a=rtpmap:") {
			rest := strings.TrimPrefix(l, "a=rtpmap:")
			parts := strings.SplitN(rest, " ", 2)
			if len(parts) == 2 && strings.HasPrefix(strings.ToLower(parts[1]), "opus/") {
				sections[cur].opusPT = parts[0]
			}
		}
	}

	out := make([]string, 0, len(lines))
	for i, l := range lines {
		section := -1
		for si, s := range sections {
			if i >= s.start && i < s.end {
				section = si
				break
			}
		}
		if section == -1 {
			out = append(out, l)
			continue
		}
		opusPT := sections[section].opusPT
		if opusPT == "" {
			out = append(out, l)
			continue
		}
		if strings.HasPrefix(l, "m=audio") {
			fields := strings.Fields(l)
			if len(fields) >= 4 {
				out = append(out, strings.Join(append(fields[:3], opusPT), " "))
			} else {
				out = append(out, l)
			}
			continue
		}
		if isPayloadAttr(l) {
			pt := payloadTypeOf(l)
			if pt != "" && pt != opusPT {
				continue
			}
		}
		out = append(out, l)
	}
	return strings.Join(out, "\r\n")
}

func isPayloadAttr(line string) bool {
	return strings.HasPrefix(line, "a=rtpmap:") ||
		strings.HasPrefix(line, "a=fmtp:") ||
		strings.HasPrefix(line, "a=rtcp-fb:")
}

func extractOpusChannels(sdp string) uint16 {
	for _, line := range strings.Split(sdp, "\r\n") {
		if !strings.HasPrefix(line, "a=rtpmap:") {
			continue
		}
		rest := strings.TrimPrefix(line, "a=rtpmap:")
		parts := strings.SplitN(rest, " ", 2)
		if len(parts) != 2 {
			continue
		}
		spec := strings.ToLower(parts[1])
		if !strings.HasPrefix(spec, "opus/") {
			continue
		}

		segs := strings.Split(spec, "/")
		if len(segs) >= 3 {
			switch segs[2] {
			case "1":
				return 1
			case "2":
				return 2
			}
		}
		return 1
	}
	return 1
}

func payloadTypeOf(line string) string {
	idx := strings.Index(line, ":")
	if idx < 0 {
		return ""
	}
	rest := line[idx+1:]
	sp := strings.Index(rest, " ")
	if sp < 0 {
		return rest
	}
	return rest[:sp]
}

File: rust/src/realtime/sdp_filter.rs
use std::collections::HashSet;

pub fn normalize_offer(sdp: &str) -> String {
    let stage_one = unify_ice_credentials(sdp);
    filter_audio_to_opus(&stage_one)
}

fn unify_ice_credentials(sdp: &str) -> String {
    let mut canonical_ufrag: Option<String> = None;
    let mut canonical_pwd: Option<String> = None;
    let mut out = String::with_capacity(sdp.len());

    for line in sdp.split_inclusive('\n') {
        let trimmed = line.trim_end_matches(['\r', '\n']);
        if let Some(rest) = trimmed.strip_prefix("a=ice-ufrag:") {
            match &canonical_ufrag {
                None => {
                    canonical_ufrag = Some(rest.to_string());
                    out.push_str(line);
                }
                Some(canonical) => {
                    out.push_str("a=ice-ufrag:");
                    out.push_str(canonical);
                    out.push_str(line_terminator(line));
                }
            }
        } else if let Some(rest) = trimmed.strip_prefix("a=ice-pwd:") {
            match &canonical_pwd {
                None => {
                    canonical_pwd = Some(rest.to_string());
                    out.push_str(line);
                }
                Some(canonical) => {
                    out.push_str("a=ice-pwd:");
                    out.push_str(canonical);
                    out.push_str(line_terminator(line));
                }
            }
        } else {
            out.push_str(line);
        }
    }
    out
}

fn filter_audio_to_opus(sdp: &str) -> String {
    let lines: Vec<&str> = sdp.split_inclusive('\n').collect();

    let mut sections: Vec<std::collections::HashMap<String, String>> = Vec::new();
    let mut cur_audio: Option<usize> = None;
    for line in &lines {
        let trimmed = line.trim_end_matches(['\r', '\n']);
        if let Some(rest) = trimmed.strip_prefix("m=") {
            if rest.starts_with("audio ") {
                sections.push(std::collections::HashMap::new());
                cur_audio = Some(sections.len() - 1);
            } else {
                cur_audio = None;
            }
            continue;
        }
        if let (Some(idx), Some(rest)) = (cur_audio, trimmed.strip_prefix("a=rtpmap:")) {
            if let Some((pt, codec_info)) = rest.split_once(' ') {
                let codec = codec_info
                    .split('/')
                    .next()
                    .unwrap_or("")
                    .to_ascii_lowercase();
                sections[idx].insert(pt.to_string(), codec);
            }
        }
    }

    let keep_codecs: HashSet<&str> = ["opus", "telephone-event", "cn"].iter().copied().collect();
    let kept_per_section: Vec<HashSet<String>> = sections
        .iter()
        .map(|map| {
            map.iter()
                .filter_map(|(pt, codec)| {
                    if keep_codecs.contains(codec.as_str()) {
                        Some(pt.clone())
                    } else {
                        None
                    }
                })
                .collect()
        })
        .collect();

    let mut out = String::with_capacity(sdp.len());
    let mut cur_audio: Option<usize> = None;
    let mut audio_count_so_far: usize = 0;
    for line in &lines {
        let trimmed = line.trim_end_matches(['\r', '\n']);
        let term = line_terminator(line);

        if let Some(rest) = trimmed.strip_prefix("m=") {
            if rest.starts_with("audio ") {
                let idx = audio_count_so_far;
                audio_count_so_far += 1;
                cur_audio = Some(idx);
                let kept = &kept_per_section[idx];

                if kept.is_empty() {
                    out.push_str(line);
                    continue;
                }

                let parts: Vec<&str> = rest.split_whitespace().collect();
                if parts.len() < 4 {
                    out.push_str(line);
                    continue;
                }
                let prefix = &parts[..3];
                let payloads = &parts[3..];
                let filtered: Vec<&str> = payloads
                    .iter()
                    .copied()
                    .filter(|pt| kept.contains(*pt))
                    .collect();
                if filtered.is_empty() {
                    out.push_str(line);
                    continue;
                }
                out.push_str("m=");
                out.push_str(prefix[0]);
                out.push(' ');
                out.push_str(prefix[1]);
                out.push(' ');
                out.push_str(prefix[2]);
                for pt in &filtered {
                    out.push(' ');
                    out.push_str(pt);
                }
                out.push_str(term);
                continue;
            } else {
                cur_audio = None;
                out.push_str(line);
                continue;
            }
        }

        if let Some(idx) = cur_audio {
            let kept = &kept_per_section[idx];
            if let Some(pt) = leading_payload_type(trimmed, "a=rtpmap:")
                .or_else(|| leading_payload_type(trimmed, "a=fmtp:"))
                .or_else(|| leading_payload_type(trimmed, "a=rtcp-fb:"))
            {
                if !kept.is_empty() && !kept.contains(&pt) {
                    continue;
                }
            }
        }

        out.push_str(line);
    }
    out
}

fn leading_payload_type(trimmed: &str, attr_prefix: &str) -> Option<String> {
    let rest = trimmed.strip_prefix(attr_prefix)?;

    let pt_token = rest.split_whitespace().next()?;
    if pt_token.chars().all(|c| c.is_ascii_digit()) {
        Some(pt_token.to_string())
    } else {
        None
    }
}

fn line_terminator(line: &str) -> &'static str {
    if line.ends_with("\r\n") {
        "\r\n"
    } else if line.ends_with('\n') {
        "\n"
    } else {
        ""
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn unifies_ice_creds_across_bundle() {
        let input = "v=0\r\n\
                     a=group:BUNDLE 0 1\r\n\
                     m=audio 1 UDP/TLS/RTP/SAVPF 96\r\n\
                     a=mid:0\r\n\
                     a=ice-ufrag:AAA1\r\n\
                     a=ice-pwd:pwd-aaaa\r\n\
                     a=rtpmap:96 opus/48000/2\r\n\
                     m=application 1 DTLS/SCTP 5000\r\n\
                     a=mid:1\r\n\
                     a=ice-ufrag:BBB2\r\n\
                     a=ice-pwd:pwd-bbbb\r\n";
        let out = normalize_offer(input);
        let ufrags: Vec<&str> = out
            .lines()
            .filter(|l| l.starts_with("a=ice-ufrag"))
            .collect();
        let pwds: Vec<&str> = out.lines().filter(|l| l.starts_with("a=ice-pwd")).collect();
        assert_eq!(ufrags, vec!["a=ice-ufrag:AAA1", "a=ice-ufrag:AAA1"]);
        assert_eq!(pwds, vec!["a=ice-pwd:pwd-aaaa", "a=ice-pwd:pwd-aaaa"]);
    }

    #[test]
    fn strips_pcmu_when_offer_lists_it_first() {
        let input = "v=0\r\n\
                     m=audio 9 UDP/TLS/RTP/SAVPF 0 96 8\r\n\
                     a=mid:0\r\n\
                     a=rtpmap:0 PCMU/8000\r\n\
                     a=rtpmap:96 opus/48000/2\r\n\
                     a=fmtp:96 minptime=10;useinbandfec=1\r\n\
                     a=rtpmap:8 PCMA/8000\r\n";
        let out = normalize_offer(input);

        let m_line = out
            .lines()
            .find(|l| l.starts_with("m=audio"))
            .expect("m=audio line present");
        let payloads: Vec<&str> = m_line.split_whitespace().skip(3).collect();
        assert_eq!(payloads, vec!["96"], "expected opus-only m-line: {m_line}");

        assert!(!out.contains("PCMU"), "PCMU should be stripped: {out}");
        assert!(!out.contains("PCMA"), "PCMA should be stripped: {out}");

        assert!(out.contains("a=rtpmap:96 opus/48000/2"));
        assert!(out.contains("a=fmtp:96 minptime=10;useinbandfec=1"));
    }

    #[test]
    fn preserves_telephone_event_and_cn() {
        let input = "v=0\r\n\
                     m=audio 9 UDP/TLS/RTP/SAVPF 96 0 101 13\r\n\
                     a=mid:0\r\n\
                     a=rtpmap:96 opus/48000/2\r\n\
                     a=rtpmap:0 PCMU/8000\r\n\
                     a=rtpmap:101 telephone-event/8000\r\n\
                     a=rtpmap:13 CN/8000\r\n";
        let out = normalize_offer(input);
        let m_line = out
            .lines()
            .find(|l| l.starts_with("m=audio"))
            .expect("m=audio line present");
        let payloads: Vec<&str> = m_line.split_whitespace().skip(3).collect();

        assert_eq!(payloads, vec!["96", "101", "13"], "got: {m_line}");
    }

    #[test]
    fn leaves_video_section_alone() {
        let input = "v=0\r\n\
                     m=video 9 UDP/TLS/RTP/SAVPF 96 97\r\n\
                     a=mid:0\r\n\
                     a=rtpmap:96 VP8/90000\r\n\
                     a=rtpmap:97 H264/90000\r\n";
        let out = normalize_offer(input);
        assert!(out.contains("m=video 9 UDP/TLS/RTP/SAVPF 96 97"));
        assert!(out.contains("a=rtpmap:96 VP8/90000"));
        assert!(out.contains("a=rtpmap:97 H264/90000"));
    }

    #[test]
    fn passthrough_when_no_opus_present() {
        let input = "v=0\r\n\
                     m=audio 9 UDP/TLS/RTP/SAVPF 0 8\r\n\
                     a=mid:0\r\n\
                     a=rtpmap:0 PCMU/8000\r\n\
                     a=rtpmap:8 PCMA/8000\r\n";
        let out = normalize_offer(input);
        assert!(out.contains("m=audio 9 UDP/TLS/RTP/SAVPF 0 8"));
        assert!(out.contains("a=rtpmap:0 PCMU/8000"));
    }
}
